src/core/smart_thread.hpp

```diff
@@ -0,0 +1,25 @@
+#pragma once
+
+#include <thread>
+
```

Comment:
This code is fairly advanced, but basically it wraps an std::thread and calls join in the destructor. When a joinable thread goes out of scope, this calls std::terminate(), this wrapper class prevents crashes by enforcing that all threads join before going out of scope.